include/zephyr/bluetooth/cs.h

```diff
@@ -540,6 +590,35 @@ struct bt_le_cs_subevent_step {
 	const uint8_t *data;
 };
 
+/** Sign-extended IQ sample extracted from step data. */
+struct bt_le_cs_iq_sample {
+	int16_t i;
+	int16_t q;
+};
+
+/** @brief Extract IQ sample from HCI-formatted PCT.
+ *
+ * Convenience function for processing 24-bit phase correction terms found
+ * in CS step data. The 12-bit signed real and imaginary components are
+ * converted to host endianness and sign-extended.
+ *
+ * @param pct 24-bit little-endian phase correction term.
+ *
+ * @return Real and imaginary terms as int16_t.
+ */
+static inline struct bt_le_cs_iq_sample bt_le_cs_parse_pct(const uint8_t pct[3])
+{
+	uint32_t sample = sys_get_le24(pct);
+	int16_t i;
+	int16_t q;
+	struct {
+		int16_t x: 12;
```

Comment:
an alternative is maybe

int16_t i = ((sample & BT_HCI_LE_CS_PCT_MASK) ^ BIT(11)) - BIT(11);
int16_t q = ((sample >> 12) ^ BIT(11)) - BIT(11);

Reply:
Both look quite hard to parse for me, however I'd at least add a code comment with a reference to the sign extension web page.

Reply:
Assigning a value to a signed integer variable that is larger than the variable can represent is technically implementation specific behavior in C.

It's probably OK and left implementation specific just in case the code runs a machine with non-twos complement integers. Maybe we can get a guarantee from Zephyr toolchain maintainers that this trick is supported?

Still, performing the conversion arithmetically avoids this caveat. I think it can be make quite clean:

```c
struct bt_le_cs_iq_sample bt_le_cs_parse_pct(const uint8_t pct[3])
{
    uint32_t sample = sys_get_le24(pct);

    /* Exract fields i and q. */
    uint16_t i2 = sample % (1<<12);
    uint16_t q2 = sample / (1<<12);

    /* Decode the fields from twos complement (bit width 12) into signed integers. */
    int16_t i = (i2 ^ (1<<11)) - (1<<11);
    int16_t q = (q2 ^ (1<<11)) - (1<<11);

    return (struct bt_le_cs_iq_sample){.i = i, .q = q};
}
```

Reply:
decided to introduce this later in a separate PR